### PR TITLE
listen on other channels when waiting for batcher

### PR DESCRIPTION
**Pull request by HagarMeir**
Signed-off-by: Hagar Meir <hagar.meir@ibm.com>

### CODE REVIEW

**Review by yacovm — CHANGES_REQUESTED** [review]
Well, the test passes for me even without the code change... 

```
yacovm@~/gopath/src/github.com/SmartBFT-Go/consensus (HagarMeir-syncLeaderAfterRestart) $ git log -n2
commit c67055114f1303cbe0ef38bd1288b28d9610ce34 (HEAD -> HagarMeir-syncLeaderAfterRestart)
Author: Hagar Meir <hagar.meir@ibm.com>
Date:   Wed Sep 4 14:22:18 2019 +0300

    listen on other channels when waiting for batcher
    
    Signed-off-by: Hagar Meir <hagar.meir@ibm.com>
    
    fix leader catch up test
    
    Signed-off-by: Hagar Meir <hagar.meir@ibm.com>

commit fc12c4912badc49ffff84c8c89224400f1680b91 (upstream/master, origin/master, origin/HEAD, master, dontprintproplsalinLogs)
Merge: 0d48447 1311288
Author: HagarMeir <hagar.meir@ibm.com>
Date:   Mon Sep 2 14:00:45 2019 +0300

    Merge pull request #318 from HagarMeir/complainWithViewNum
    
    ignore old complains
yacovm@~/gopath/src/github.com/SmartBFT-Go/consensus (HagarMeir-syncLeaderAfterRestart) $ git diff HEAD~1 test/basic_test.go | wc -l
68
yacovm@~/gopath/src/github.com/SmartBFT-Go/consensus (HagarMeir-syncLeaderAfterRestart) $ git diff HEAD~1 internal/bft/controller.go | wc -l
0
yacovm@~/gopath/src/github.com/SmartBFT-Go/consensus (HagarMeir-syncLeaderAfterRestart) $ 

yacovm@~/gopath/src/github.com/SmartBFT-Go/consensus (HagarMeir-syncLeaderAfterRestart) $ go test -count 1 ./test/... 
ok  	github.com/SmartBFT-Go/consensus/test	12.525s

```

File: test/basic_test.go [hunk 1]
```diff
@@ -540,44 +540,18 @@ func TestLeaderCatchingUpAfterViewChange(t *testing.T) {
...
+		n1.Submit(Request{ID: fmt.Sprintf("%d", reqID), ClientID: "alice"})
+		n2.Submit(Request{ID: fmt.Sprintf("%d", reqID), ClientID: "alice"})
+		<-n1.Delivered // Wait for new leader to commit
+		<-n2.Delivered // Wait for follower to commit
+		caughtUp := waitForCatchup(reqID, n0.Delivered)
+		if caughtUp {
+			return
+		}
+	}
+	t.Log("Didn't catch up")
```
Comment (yacovm) [inline test/basic_test.go]: Fatalf

**Review by yacovm — CHANGES_REQUESTED** [review]
I think that the below solution is slightly easier to reason about and has a lower surface area:

```
diff --git a/internal/bft/batcher.go b/internal/bft/batcher.go
index ae8c20d..4001562 100644
--- a/internal/bft/batcher.go
+++ b/internal/bft/batcher.go
@@ -83,6 +83,15 @@ func (b *BatchBuilder) Close() {
 	close(b.closeChan)
 }
 
+func (b *BatchBuilder) Closed() bool {
+	select {
+	case <-b.closeChan:
+		return true
+	default:
+		return false
+	}
+}
+
 // Reset resets the remainder and reopens the close channel to allow calling NextBatch
 func (b *BatchBuilder) Reset() {
 	b.closeLock.Lock()
diff --git a/internal/bft/controller.go b/internal/bft/controller.go
index 1a60f2e..cf463ee 100644
--- a/internal/bft/controller.go
+++ b/internal/bft/controller.go
@@ -30,6 +30,7 @@ type Batcher interface {
 	BatchRemainder(remainder [][]byte)
 	PopRemainder() [][]byte
 	Close()
+	Closed() bool
 	Reset()
 }
 
@@ -304,6 +305,9 @@ func (c *Controller) abortView() {
 }
 
 func (c *Controller) Sync() {
+	if iAmLeader, _ := c.iAmTheLeader(); iAmLeader {
+		c.Batcher.Close()
+	}
 	c.grabSyncToken()
 }
 
@@ -330,7 +334,7 @@ func (c *Controller) getNextBatch() [][]byte {
 	var validRequests [][]byte
 	for len(validRequests) == 0 { // no valid requests in this batch
 		requests := c.Batcher.NextBatch()
-		if c.stopped() {
+		if c.stopped() || c.Batcher.Closed() {
 			return nil
 		}
 		for _, req := range requests {
@@ -470,7 +474,7 @@ func (c *Controller) relinquishLeaderToken() {
 func (c *Controller) Start(startViewNumber uint64, startProposalSequence uint64) {
 	c.controllerDone.Add(1)
 	c.stopOnce = sync.Once{}
-	c.syncChan = make(chan struct{})
+	c.syncChan = make(chan struct{}, 1)
 	c.stopChan = make(chan struct{})
 	c.leaderToken = make(chan struct{}, 1)
 	c.decisionChan = make(chan decision)
diff --git a/internal/bft/controller_test.go b/internal/bft/controller_test.go
index a1cedc9..90ac08e 100644
--- a/internal/bft/controller_test.go
+++ b/internal/bft/controller_test.go
@@ -72,6 +72,7 @@ func TestControllerLeaderBasic(t *testing.T) {
 	log := basicLog.Sugar()
 	batcher := &mocks.Batcher{}
 	batcher.On("Close")
+	batcher.On("Closed").Return(false)
 	batcherChan := make(chan struct{})
 	var once sync.Once
 	batcher.On("NextBatch").Run(func(args mock.Arguments) {
@@ -111,6 +112,7 @@ func TestLeaderPropose(t *testing.T) {
 	req := []byte{1}
 	batcher := &mocks.Batcher{}
 	batcher.On("Close")
+	batcher.On("Closed").Return(false)
 	batcher.On("NextBatch").Return([][]byte{req}).Once()
 	batcher.On("NextBatch").Return([][]byte{req}).Once()
 	batcher.On("PopRemainder").Return([][]byte{})
@@ -227,6 +229,7 @@ func TestViewChanged(t *testing.T) {
 	req := []byte{1}
 	batcher := &mocks.Batcher{}
 	batcher.On("Close")
+	batcher.On("Closed").Return(false)
 	batcher.On("Reset")
 	batcher.On("NextBatch").Return([][]byte{req})
 	verifier := &mocks.VerifierMock{}
@@ -416,6 +419,7 @@ func TestSyncInform(t *testing.T) {
 	req := []byte{1}
 	batcher := &mocks.Batcher{}
 	batcher.On("Close")
+	batcher.On("Closed").Return(false)
 	batcher.On("Reset")
 	batcher.On("NextBatch").Return([][]byte{req})
 	verifier := &mocks.VerifierMock{}
diff --git a/internal/bft/mocks/batcher.go b/internal/bft/mocks/batcher.go
index 0026b21..0825f1e 100644
--- a/internal/bft/mocks/batcher.go
+++ b/internal/bft/mocks/batcher.go
@@ -19,6 +19,20 @@ func (_m *Batcher) Close() {
 	_m.Called()
 }
 
+// Closed provides a mock function with given fields:
+func (_m *Batcher) Closed() bool {
+	ret := _m.Called()
+
+	var r0 bool
+	if rf, ok := ret.Get(0).(func() bool); ok {
+		r0 = rf()
+	} else {
+		r0 = ret.Get(0).(bool)
+	}
+
+	return r0
+}
+
 // NextBatch provides a mock function with given fields:
 func (_m *Batcher) NextBatch() [][]byte {
 	ret := _m.Called()

```

**Comment by HagarMeir** [comment]
> Well, the test passes for me even without the code change...

The test is flaky without the change

**Pull request closed without merging**